File: scripts/import-ao-zip.py
#!/usr/bin/env python3
"""Import AO documents from zip without full extraction (disk-safe)."""
from __future__ import annotations

import json
import os
import re
import sys
import time
import urllib.error
import urllib.request
import zipfile
from pathlib import PurePosixPath

ZIP_PATH = sys.argv[1] if len(sys.argv) > 1 else "/Users/mathieufaessel/Downloads/OneDrive_1_30-08-2026.zip"
MAX_BYTES = 50 * 1024 * 1024
SUPABASE_URL = os.environ.get("SUPABASE_URL") or os.environ.get("VITE_SUPABASE_URL", "")
SERVICE_KEY = os.environ.get("SUPABASE_SERVICE_ROLE_KEY", "")

AO_MAP = [
    ("260702_Arc de triomphe", "260702"),
    ("260713_12H_GOUSSAINVILLE", "260713"),
    (
        "Exploitation, maintenance préventive et corrective des installations d’éclairage et d’électricité de l’Arc de Triomphe",
        "26-641",
    ),
    ("MAIRIE DE JOUARS PONTCHARTRAIN", "AO-JOUARS"),
    ("MAIRIE DE OSNY", "AO-OSNY"),
    ("Mairie d'Evry-Courcouronnes", "AO-EVRY"),
    ("Mairie de champagne sur oise", "AO-CHAMP"),
    ("rue flachat Asniére sur seine", "AO-ASN"),
]


def api(method: str, path: str, data: bytes | None = None, content_type: str = "application/json", extra: dict | None = None):
    url = f"{SUPABASE_URL.rstrip('/')}{path}"
    headers = {
        "apikey": SERVICE_KEY,
        "Authorization": f"Bearer {SERVICE_KEY}",
    }
    if content_type:
        headers["Content-Type"] = content_type
    if extra:
        headers.update(extra)
    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=120) as res:
            body = res.read()
            if not body:
                return None
            return json.loads(body)
    except urllib.error.HTTPError as e:
        raise RuntimeError(f"{e.code} {path}: {e.read().decode()[:300]}") from e


def doc_type(path: str) -> str:
    lower = path.lower()
    if "/00_dce/" in lower or "\\00_dce\\" in lower:
        if "rc_" in lower or "reglement" in lower:
            return "rc"
        if "cctp" in lower:
            return "cctp"
        if "ccap" in lower:
            return "ae"
        if "dpgf" in lower:
            return "dpgf"
        if "bpu" in lower:
            return "bpu"
        return "dce"
    if "memoire" in lower or "m+moire" in lower or "mémoire" in lower:
        return "memoire"
    if "/03_" in lower or "reponse" in lower or "réponse" in lower:
        return "reponse"
    if "attestationdepot" in lower:
        return "reponse"
    if "chiffrage" in lower:
        return "annexe"
    return "annexe"


def should_skip(name: str) -> bool:
    base = PurePosixPath(name).name
    if base.startswith("~$") or base.startswith("._") or base == ".DS_Store":
        return True
    if "__MACOSX" in name:
        return True
    if base.endswith(".tmp"):
        return True
    if base.lower().endswith(".zip"):
        return True
    return False


def fmt_size(n: int) -> str:
    if n < 1024 * 1024:
        return f"{n / 1024:.0f} Ko"
    return f"{n / (1024 * 1024):.1f} Mo"


def sanitize(name: str) -> str:
    return re.sub(r"[^a-zA-Z0-9._-]", "_", name)


def folder_match(entry: str, prefix: str) -> bool:
    parts = PurePosixPath(entry).parts
    if not parts:
        return False
    root = parts[0]
    return root == prefix or root.lower().strip() == prefix.lower().strip()


def main():
    if not SUPABASE_URL or not SERVICE_KEY:
        print("Missing SUPABASE_URL / SUPABASE_SERVICE_ROLE_KEY", file=sys.stderr)
        sys.exit(1)

    print(f"[1/2] Lecture zip {ZIP_PATH}…")
    aos = api("GET", "/rest/v1/appels_offres?select=id,reference")
    ao_by_ref = {a["reference"]: a["id"] for a in aos}

    uploaded = skipped = skipped_large = 0

    with zipfile.ZipFile(ZIP_PATH, "r") as zf:
        names = [n for n in zf.namelist() if not n.endswith("/")]
        print(f"[2/2] {len(names)} entrées — upload (< 50 Mo)…\n")

        for folder_prefix, reference in AO_MAP:
            ao_id = ao_by_ref.get(reference)
            if not ao_id:
                print(f"⚠ AO manquant: {reference}")
                continue

            ao_count = 0
            for entry in names:
                if not folder_match(entry, folder_prefix):
                    continue
                if should_skip(entry):
                    skipped += 1
                    continue

                info = zf.getinfo(entry)
                if info.file_size > MAX_BYTES:
                    print(f"  ⊘ {PurePosixPath(entry).name} ({fmt_size(info.file_size)} > 50 Mo)")
                    skipped_large += 1
                    skipped += 1
                    continue

                file_name = PurePosixPath(entry).name
                dtype = doc_type(entry)
                storage_path = f"{ao_id}/{dtype}/{int(time.time() * 1000)}_{sanitize(file_name)}"

                try:
                    data = zf.read(entry)
                    encoded = "/".join(urllib.request.quote(p) for p in storage_path.split("/"))
                    api(
                        "POST",
                        f"/storage/v1/object/ao-documents/{encoded}",
                        data=data,
                        content_type="application/octet-stream",
                        extra={"x-upsert": "false"},
                    )
                    api(
                        "POST",
                        "/rest/v1/ao_documents",
                        data=json.dumps(
                            {
                                "ao_id": ao_id,
                                "type": dtype,
                                "nom_fichier": file_name,
                                "storage_path": storage_path,
                                "fichier_url": storage_path,
                                "taille_octets": len(data),
                                "notes": "Import OneDrive 30/08/2026",
                                "uploaded_by_email": "import@meselec.fr",
                            }
                        ).encode(),
                        extra={"Prefer": "return=minimal"},
                    )
                    uploaded += 1
                    ao_count += 1
                    print(f"  ✓ {file_name} ({fmt_size(len(data))})")
                except Exception as e:
                    print(f"  ✗ {file_name}: {e}")
                    skipped += 1

            print(f"→ {reference}: {ao_count} fichier(s)\n")

    print(
        f"Terminé — {uploaded} uploadés, {skipped_large} ignorés (> 50 Mo), {skipped - skipped_large} autres ignorés"
    )


if __name__ == "__main__":
    main()

File: scripts/import-ao-zip.ts
/**
 * Importe les documents AO depuis l'archive OneDrive (AO déjà seedés en base).
 * Ignore les fichiers > 50 Mo et les archives imbriquées (.zip).
 */
import { execSync } from "node:child_process";
import { mkdtempSync, readdirSync, readFileSync, rmSync, statSync } from "node:fs";
import { basename, join, relative } from "node:path";
import { tmpdir } from "node:os";

const ZIP_PATH = process.argv[2] ?? "/Users/mathieufaessel/Downloads/OneDrive_1_30-08-2026.zip";
const MAX_BYTES = 50 * 1024 * 1024;
const SUPABASE_URL = (process.env.SUPABASE_URL ?? process.env.VITE_SUPABASE_URL ?? "").replace(
  /\/$/,
  "",
);
const SERVICE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY ?? "";

if (!SUPABASE_URL || !SERVICE_KEY) {
  console.error("Missing SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY");
  process.exit(1);
}

const BUCKET = "ao-documents";
const headers = {
  apikey: SERVICE_KEY,
  Authorization: `Bearer ${SERVICE_KEY}`,
};

const AO_MAP: Array<{ folderPrefix: string; reference: string }> = [
  { folderPrefix: "260702_Arc de triomphe", reference: "260702" },
  { folderPrefix: "260713_12H_GOUSSAINVILLE", reference: "260713" },
  {
    folderPrefix:
      "Exploitation, maintenance préventive et corrective des installations d’éclairage et d’électricité de l’Arc de Triomphe",
    reference: "26-641",
  },
  { folderPrefix: "MAIRIE DE JOUARS PONTCHARTRAIN", reference: "AO-JOUARS" },
  { folderPrefix: "MAIRIE DE OSNY", reference: "AO-OSNY" },
  { folderPrefix: "Mairie d'Evry-Courcouronnes", reference: "AO-EVRY" },
  { folderPrefix: "Mairie de champagne sur oise", reference: "AO-CHAMP" },
  { folderPrefix: "rue flachat Asniére sur seine", reference: "AO-ASN" },
];

async function rest<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${SUPABASE_URL}/rest/v1/${path}`, {
    ...init,
    headers: { ...headers, "Content-Type": "application/json", ...init?.headers },
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(`${res.status} ${path}: ${body}`);
  }
  if (res.status === 204) return undefined as T;
  return (await res.json()) as T;
}

async function uploadStorage(path: string, buffer: Buffer, contentType: string) {
  const encoded = path.split("/").map(encodeURIComponent).join("/");
  const res = await fetch(`${SUPABASE_URL}/storage/v1/object/${BUCKET}/${encoded}`, {
    method: "POST",
    headers: {
      ...headers,
      "Content-Type": contentType,
      "x-upsert": "false",
    },
    body: buffer,
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(`Storage ${res.status}: ${body}`);
  }
}

function docTypeFromPath(path: string): string {
  const lower = path.toLowerCase();
  if (lower.includes("/00_dce/") || lower.includes("\\00_dce\\")) {
    if (lower.includes("rc_") || lower.includes("reglement")) return "rc";
    if (lower.includes("cctp")) return "cctp";
    if (lower.includes("ccap")) return "ae";
    if (lower.includes("dpgf")) return "dpgf";
    if (lower.includes("bpu")) return "bpu";
    return "dce";
  }
  if (lower.includes("memoire") || lower.includes("m+moire") || lower.includes("mémoire"))
    return "memoire";
  if (lower.includes("/03_") || lower.includes("reponse") || lower.includes("réponse")) return "reponse";
  if (lower.includes("attestationdepot")) return "reponse";
  if (lower.includes("chiffrage")) return "annexe";
  return "annexe";
}

function shouldSkipPath(path: string): boolean {
  const name = basename(path);
  if (name.startsWith("~$") || name.startsWith("._") || name === ".DS_Store") return true;
  if (path.includes("__MACOSX")) return true;
  if (name.endsWith(".tmp")) return true;
  if (name.toLowerCase().endsWith(".zip")) return true;
  return false;
}

function walkFiles(dir: string): string[] {
  const out: string[] = [];
  for (const entry of readdirSync(dir)) {
    const full = join(dir, entry);
    const st = statSync(full);
    if (st.isDirectory()) out.push(...walkFiles(full));
    else out.push(full);
  }
  return out;
}

function formatSize(bytes: number): string {
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(0)} Ko`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} Mo`;
}

function resolveFolderPath(tmp: string, folderPrefix: string): string | null {
  const direct = join(tmp, folderPrefix);
  try {
    if (statSync(direct).isDirectory()) return direct;
  } catch {
    /* fuzzy */
  }
  const normalized = folderPrefix.toLowerCase().replace(/\s+/g, " ").trim();
  for (const entry of readdirSync(tmp)) {
    try {
      if (!statSync(join(tmp, entry)).isDirectory()) continue;
    } catch {
      continue;
    }
    if (entry.toLowerCase().replace(/\s+/g, " ").trim() === normalized) {
      return join(tmp, entry);
    }
  }
  return null;
}

function mimeFor(name: string): string {
  const ext = name.split(".").pop()?.toLowerCase() ?? "";
  const map: Record<string, string> = {
    pdf: "application/pdf",
    docx: "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    xlsx: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    pptx: "application/vnd.openxmlformats-officedocument.presentationml.presentation",
    jpg: "image/jpeg",
    jpeg: "image/jpeg",
    png: "image/png",
  };
  return map[ext] ?? "application/octet-stream";
}

async function uploadFile(aoId: string, absPath: string, relFromRoot: string) {
  const fileName = basename(absPath);
  const type = docTypeFromPath(relFromRoot);
  const storagePath = `${aoId}/${type}/${Date.now()}_${fileName.replace(/[^a-zA-Z0-9._-]/g, "_")}`;
  const buffer = readFileSync(absPath);

  await uploadStorage(storagePath, buffer, mimeFor(fileName));

  await rest("ao_documents", {
    method: "POST",
    body: JSON.stringify({
      ao_id: aoId,
      type,
      nom_fichier: fileName,
      storage_path: storagePath,
      fichier_url: storagePath,
      taille_octets: buffer.length,
      notes: "Import OneDrive 30/08/2026",
      uploaded_by_email: "import@meselec.fr",
    }),
    headers: { Prefer: "return=minimal" },
  });
}

async function main() {
  const tmp = mkdtempSync(join(tmpdir(), "ao-import-"));
  console.log(`[1/3] Extraction de ${ZIP_PATH}…`);
  execSync(`unzip -q "${ZIP_PATH}" -d "${tmp}"`);

  console.log("[2/3] Chargement des AO…");
  const aos = await rest<Array<{ id: string; reference: string }>>(
    "appels_offres?select=id,reference",
  );
  const aoByRef = new Map(aos.map((a) => [a.reference, a.id]));

  let uploaded = 0;
  let skipped = 0;
  let skippedLarge = 0;

  console.log("[3/3] Upload (< 50 Mo, sans .zip)…\n");

  for (const { folderPrefix, reference } of AO_MAP) {
    const aoId = aoByRef.get(reference);
    if (!aoId) {
      console.warn(`⚠ AO introuvable: ${reference}`);
      continue;
    }

    const folderPath = resolveFolderPath(tmp, folderPrefix);
    if (!folderPath) {
      console.warn(`⚠ Dossier introuvable: ${folderPrefix}`);
      continue;
    }

    let aoUploaded = 0;
    for (const file of walkFiles(folderPath)) {
      const rel = relative(folderPath, file);
      if (shouldSkipPath(rel)) {
        skipped++;
        continue;
      }

      const size = statSync(file).size;
      if (size > MAX_BYTES) {
        console.log(`  ⊘ ${basename(rel)} (${formatSize(size)} > 50 Mo)`);
        skippedLarge++;
        skipped++;
        continue;
      }

      try {
        await uploadFile(aoId, file, rel);
        uploaded++;
        aoUploaded++;
        console.log(`  ✓ ${basename(rel)} (${formatSize(size)})`);
      } catch (e) {
        console.warn(`  ✗ ${basename(rel)}:`, e instanceof Error ? e.message : e);
        skipped++;
      }
    }

    console.log(`→ ${reference}: ${aoUploaded} fichier(s)\n`);
  }

  rmSync(tmp, { recursive: true, force: true });
  console.log(
    `Terminé — ${uploaded} uploadés, ${skippedLarge} ignorés (> 50 Mo), ${skipped - skippedLarge} autres ignorés`,
  );
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
